File: Fundamentals_of_Software_Design/C_language/Stack/stack.c
#include <stdlib.h>

#include "stack.h"

#define INITIAL_CAPACITY 16

Stack* stack_create() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    if (!stack) return NULL;
    stack->data = (Pointer*)malloc(INITIAL_CAPACITY * sizeof(Pointer));
    if (!stack->data) {
        free(stack);
        return NULL;
    }
    stack->size = 0;
    stack->capacity = INITIAL_CAPACITY;
    return stack;
}

void stack_free(Stack* stack) {
    if (!stack) return;
    free(stack->data);
    free(stack);
}

void stack_push(Stack* stack, Pointer value) {
    if (!stack) return;
    if (stack->size >= stack->capacity) {
        size_t new_capacity = stack->capacity * 2;
        Pointer* new_data = (Pointer*)realloc(stack->data, new_capacity * sizeof(Pointer));
        if (!new_data) return;
        stack->data = new_data;
        stack->capacity = new_capacity;
    }
    stack->data[stack->size++] = value;
}

Pointer stack_pop(Stack* stack) {
    if (!stack || stack->size == 0) return NULL;
    return stack->data[--stack->size];
}

Pointer stack_peek(Stack* stack) {
    if (!stack || stack->size == 0) return NULL;
    return stack->data[stack->size - 1];
}

int stack_empty(Stack* stack) {
    return !stack || stack->size == 0;
}

size_t stack_size(Stack* stack) {
    return stack ? stack->size : 0;
}

void stack_clear(Stack* stack) {
    if (stack) stack->size = 0;
}

File: Fundamentals_of_Software_Design/Python_language/Stack/stack.py
INITIAL_CAPACITY = 16

class Stack:
    __slots__ = ('data', 'size', 'capacity')
    def __init__(self):
        self.data = [None] * INITIAL_CAPACITY
        self.size = 0
        self.capacity = INITIAL_CAPACITY

def stack_create():
    return Stack()

def stack_free(stack):
    stack.data = None
    stack.size = 0
    stack.capacity = 0

def stack_push(stack, value):
    if stack.size >= stack.capacity:
        new_capacity = stack.capacity * 2
        new_data = [None] * new_capacity
        for i in range(stack.size):
            new_data[i] = stack.data[i]
        stack.data = new_data
        stack.capacity = new_capacity
    stack.data[stack.size] = value
    stack.size += 1

def stack_pop(stack):
    if stack.size == 0:
        return None
    stack.size -= 1
    return stack.data[stack.size]

def stack_peek(stack):
    if stack.size == 0:
        return None
    return stack.data[stack.size - 1]

def stack_empty(stack):
    return stack.size == 0

def stack_size(stack):
    return stack.size

def stack_clear(stack):
    stack.size = 0
